test/functional/ui/input_spec.lua

```diff
@@ -126,3 +128,208 @@ describe('input utf sequences that contain CSI/K_SPECIAL', function()
+      l  Y           *@F
+      l  Z           *@:
+      l  [           *@/
+      l  \           *@\
+      l  ]           *@=
+      l  _           *@{
+      l  a           *@a
+      l  b           *@x
+      l  c           *@j
+      l  d           *@e
+      l  e           *@.
+      l  f           *@u
+      l  g           *@i
+      l  h           *@d
+      l  i           *@c
+      l  j           *@h
+      l  k           *@t
+      l  l           *@n
+      l  m           *@m
+      l  n           *@b
+      l  o           *@r
+      l  p           *@l
+      l  q           *@'
+      l  r           *@p
+      l  s           *@o
+      l  t           *@y
+      l  u           *@g
+      l  v           *@k
+      l  w           *@,
+      l  x           *@q
+      l  y           *@f
+      l  z           *@;
+      l  {           *@?
+      l  |           *@|
+      l  }           *@+]]))
+    end)
+    it("'keymap' mappings are not remappable", function()
+    end)
+  end)
+  describe("'iminsert' option", function()
+    it("Uses :lmap in insert mode when ON", function()
+      feed('il<esc>')
+      expect('alllaaa')
+    end)
+    it("Ignores :lmap in insert mode when OFF", function()
+      feed_command('set iminsert=0')
+      feed('il<esc>')
+      expect('llllaaa')
+    end)
+    it("Can be toggled with <C-^> in insert mode", function()
+      feed('i<C-^>l<C-^>l<esc>')
+      expect('lalllaaa')
+      eq(eval('&iminsert'), 1)
+      feed('i<C-^><esc>')
+      eq(eval('&iminsert'), 0)
+    end)
+  end)
+  describe("'imsearch' option", function()
+    it("Uses :lmap at search prompt when ON", function()
+      feed('/lll<cr>3x')
+      expect('lll')
+    end)
+    it("Ignores :lmap at search prompt when OFF", function()
+      feed_command('set imsearch=0')
+      feed('gg/lll<cr>3x')
+      expect('aaa')
+    end)
+    it("Can be toggled with C-^", function()
+      eq(eval('&imsearch'), 1)
+      feed('/<C-^>lll<cr>3x')
+      expect('aaa')
+      eq(eval('&imsearch'), 0)
+      feed('u0/<C-^>lll<cr>3x')
+      expect('lll')
+      eq(eval('&imsearch'), 1)
+    end)
+    it("can follow 'iminsert'", function()
+      feed_command('set imsearch=-1')
+      feed('/lll<cr>3x')
+      expect('lll')
+      eq(eval('&imsearch'), -1)
+      eq(eval('&iminsert'), 1)
+      feed('u/<C-^>lll<cr>3x')
+      expect('aaa')
+      eq(eval('&imsearch'), -1)
+      eq(eval('&iminsert'), 0)
+    end)
+  end)
+  it(":lmap not applied to macros", function()
+    command("call setreg('a', 'il')")
+    feed('@a')
+    expect('llllaaa')
+    eq(call('getreg', 'a'), 'il')
+  end)
+  it(":lmap applied to macro recording", function()
+    feed('qail<esc>q@a')
+    expect('aalllaaa')
+    eq(call('getreg', 'a'), 'ia')
+  end)
+  it(":lmap not applied to mappings", function()
+    feed_command('imap t l')
+    feed('it<esc>')
+    expect('llllaaa')
+  end)
+  it("mappings applied to keys created with :lmap", function()
+    feed_command('imap a x')
```

Comment:
in general use command() instead of feed_command() unless the asynchrony is intended. Also feed_command() doesn't bubble errors up (though it does set v:errmsg, but that needs to be checked explicitly)